controller interface

b9a9s9 opened the pull request:
@gerritvanos 
In this release the following code can be found:
- An interface for different types of controllers
- A dummy using the interface
- Catch2 tests for the interface

The UML and information about the interface can be found at:
https://github.com/R2D2-2019/R2D2-2019/wiki/Manual-Control#uml-diagrams

b9a9s9 committed:
initial commit

b9a9s9 committed:
controller_interface_c implemented. still needs testing

b9a9s9 committed:
controller dummy added

b9a9s9 committed:
processed feedback:
- R2D2 namespace changed to r2d2
- documentary moved to hpp and expanded.

b9a9s9 committed:
namespace fix from R2D2 to r2d2 in main.

b9a9s9 committed:
Merge pull request #4 from R2D2-2019/feature-interface

Feature interface

b9a9s9 committed:
Catch2 tests for dummy/interface.

b9a9s9 committed:
Merge pull request #5 from R2D2-2019/feature-interface

Catch2 tests for dummy/interface.

itzandroidtab commented on .travis.yml:
```diff
@@ -0,0 +1,78 @@
+dist: trusty
```
Maybe a update to the new travis.yml is a good idea

itzandroidtab commented on code/headers/controller_interface_c.hpp:
```diff
@@ -0,0 +1,43 @@
+    enum joysticks {joystick_l, joystick_r};
+
+
+/**
```
The indentation for the documentation should be the same as the class.

itzandroidtab commented on code/headers/dummy_controller_c.hpp:
```diff
@@ -0,0 +1,74 @@
+    
+    
+
+    class dummy_controller_c: public controller_interface_c{
```
Missing space before '{'

itzandroidtab commented on code/headers/dummy_controller_c.hpp:
```diff
@@ -0,0 +1,74 @@
+    class dummy_controller_c: public controller_interface_c{
+    private:
+    //initializing all buttons, sliders and joysticks at their base (0)
+    bool buttons[12] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
```
Is it not cleaner to use:

```cpp
bool buttons[12] = {};
```
?

b9a9s9 replied:
I didn't know that initializing like that garanties that all the values are 0. I will change this in the next commit

itzandroidtab commented on code/main.cpp:
```diff
@@ -0,0 +1,12 @@
+    WDT->WDT_MR = WDT_MR_WDDIS;
+    hwlib::wait_ms(1000);
+      hwlib::cout << "this works via arduino";
+      hwlib::wait_ms(1000);
```
Could you correctly indent these lines?

itzandroidtab commented on test/main.cpp:
```diff
@@ -0,0 +1,4 @@
+#include "ostream"
```
Can you change the include to 

`<>`

itzandroidtab commented on code/headers/controller_interface_c.hpp:
```diff
@@ -0,0 +1,43 @@
+
+namespace r2d2::manual_control {
+    
+    enum buttons {button_a, button_b, button_x, button_y, dpad_up, dpad_down, dpad_left, dpad_right, bumper_left, bumper_right, menu_left, menu_right};
```
This line is way to long. You should format this file. As the indents in this file are strange as well.

itzandroidtab commented on code/headers/dummy_controller_c.hpp:
```diff
@@ -0,0 +1,74 @@
+namespace r2d2::manual_control {
+    
+    
+
```
Can you remove some of these newlines?

itzandroidtab commented:
Most of these files are not formatted. As most indents are not correct and the length of some lines is to long. Can you format the files and ill take a look again.

b9a9s9 committed:
running clang

b9a9s9 committed:
Merge pull request #7 from R2D2-2019/feature-interface

running clang

b9a9s9 commented:
@itzandroidtab 
the files have been formatted

itzandroidtab commented on code/headers/joystick_value_s.hpp:
```diff
@@ -0,0 +1,4 @@
+struct joystick_value_s {
```
This header needs include guards

itzandroidtab commented on code/main.cpp:
```diff
@@ -0,0 +1,11 @@
+#include "hwlib.hpp"
```
Can you change the `""` to `<>`

itzandroidtab commented on test/main_dummy_controller_test.cpp:
```diff
@@ -0,0 +1,46 @@
+#include "ostream"
```
Can you change the "" to <>

itzandroidtab commented on code/headers/controller_interface_c.hpp:
```diff
@@ -0,0 +1,57 @@
+     *  This class is an interface for different kinds of controllers.
+     *
+     *  \details
+     *   Every controller will inherit from this class. This class is a custom
```
Nitpick `Every` doesnt allign with `This` of \brief

itzandroidtab commented on test/main_dummy_controller_test.cpp:
```diff
@@ -0,0 +1,46 @@
+#include <catch.hpp>
+
+TEST_CASE("testing buttons", " [interface]") {
+    auto dummy_controller = r2d2::manual_control::dummy_controller_c();
```
A `using namespace r2d2::manual_control` will make this file much more readable.

itzandroidtab commented on test/main_dummy_controller_test.cpp:
```diff
@@ -0,0 +1,46 @@
+    CHECK(dummy_controller.get_slider(r2d2::manual_control::slider_r) == 1);
+
+    // check joysticks
+    CHECK(dummy_controller.get_joystick(r2d2::manual_control::joystick_l).x ==
```
These check's are strangely formatted. For example you can format them like this:

```cpp
CHECK(dummy_controller.get_joystick(
    r2d2::manual_control::joystick_l
).x == 0);
```

b9a9s9 replied:
This is what clang did, but its solved now anyways since "r2d2::manual_control" is gone

itzandroidtab commented:
It looks like all the files are named wrongly. All the files should be renamed to the name without the `_c` and `_s`

itzandroidtab commented on test/main_dummy_controller_test.cpp:
```diff
@@ -0,0 +1,46 @@
+        // check buttons
+        r2d2::manual_control::buttons button =
+            static_cast<r2d2::manual_control::buttons>(x);
+        CHECK(dummy_controller.get_button(button) == 0);
```
Can you change the `CHECK()` to `REQUIRE()` as this will fail the test when the require fails instead of continuing.

b9a9s9 replied:
I read up on the difference between check and require and it said that check is used when its useful to know the result of all the tests, even if for example the first test fails. In this case i think it's useful to know the result of each test so you know which buttons, sliders and joysticks do, and do not work.

LRstudentHU reviewed (approved):
I'll approve as I have nothing to add to Koen's points. When Koen has approved, you can merge.

b9a9s9 committed:
PR feedback processed

fixed "" to <> on includes
renamed filed (removed _c and _s)
arrays initialized as array[] = {} instead of array[] = {0, 0, 0.....}
added include guards to joystick_value.hpp
main_dummy_controller_test.cpp more readable with using namespace

b9a9s9 committed:
Merge pull request #8 from R2D2-2019/feature-interface

PR feedback processed

itzandroidtab commented:
@b9a9s9 Looks good, Just 3 small points: Can you remove the `.vscode/ipch` directory and add it to the .gitignore and update the .travis.yml?

b9a9s9 committed:
Removing unnecessary files, updating travis and updating gitignore(now ignoring .vscode)

b9a9s9 committed:
Add vscode to gitignore

b9a9s9 committed:
Merge pull request #9 from R2D2-2019/feature-interface

Removing unnecessary files, updating travis and updating gitignore

b9a9s9 commented on .gitignore:
```diff
@@ -0,0 +1,15 @@
+*.lss
+*.exe
+*.lst
+*.vscode/*
```
```suggestion
.vscode/ipch
```

b9a9s9 committed:
Update gitignore

Only ignore ipch from .vscode

b9a9s9 committed:
Merge pull request #10 from R2D2-2019/feature-interface

Update gitignore
